libs-scala/scala-utils/src/main/scala/com/daml/scalautil/concurrent/FutureOf.scala

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.scalautil.concurrent
+
+import scala.language.{higherKinds, implicitConversions}
+import scala.{concurrent => sc}
+import scala.util.Try
+
+import scalaz.{Catchable, Cobind, Isomorphism, Leibniz, MonadError, Nondeterminism, Semigroup}
+import Isomorphism.<~>
+import Leibniz.===
+import scalaz.std.scalaFuture._
+
+sealed abstract class FutureOf {
+
+  /** We don't use [[sc.Future]] as the upper bound because it has methods that
+    * collide with the versions we want to use, i.e. those that preserve the
+    * phantom `EC` type parameter.  By contrast, [[sc.Awaitable]] has only the
+    * `ready` and `result` methods, which are mostly useless.
+    */
+  type T[-EC, +A] <: sc.Awaitable[A]
```

Comment:
After writing the initial implementation and looking at it for a bit, I started to worry about the placement of `EC`.  I mean to place limits on the `map`, `flatMap`, &c operations called on this Future.  But the `EC` on a Future has nothing to do with the execution context in which *that future was evaluated*, it constrains only *the combinators on that Future, i.e. other Futures it may produce*.

My worry was "this EC on the Future has nothing to do with the Future; it's not really part of it".  I thought, "should I complicate the API by adding a kind of 'projection', a method you call before map, flatmap, &c to declare an EC for the following call?"  But then I remembered

> None of this *means* anything, none of it *is* anything.  It's all uninterpreted symbols.  This, in particular, isn't even structural, never mind operational; it's only nominal, so it's even more "meaning"-less than all programs already are.

Speaking *practically*, and abandoning a quest for "meaning" in phantoms, the only thing that matters is **if this purely nominal abstraction well-types the kinds of things we want to be accepted, and ill-types the kinds of things we want to reject** (without further annotation that you know what you're doing).  Any constraints and constraint rules that make that happen are worthy of consideration.